util/network-devp2p/src/discovery.rs

```diff
@@ -457,61 +471,105 @@ impl<'a> Discovery<'a> {
 		entry.endpoint.is_allowed(&self.ip_filter) && entry.id != self.id
 	}
 
-	fn on_ping(&mut self, rlp: &Rlp, node: &NodeId, from: &SocketAddr, echo_hash: &[u8]) -> Result<Option<TableUpdates>, Error> {
+	fn on_ping(&mut self, rlp: &Rlp, node_id: &NodeId, from: &SocketAddr, echo_hash: &[u8]) -> Result<Option<TableUpdates>, Error> {
 		trace!(target: "discovery", "Got Ping from {:?}", &from);
-		let source = NodeEndpoint::from_rlp(&rlp.at(1)?)?;
-		let dest = NodeEndpoint::from_rlp(&rlp.at(2)?)?;
+		let ping_from = NodeEndpoint::from_rlp(&rlp.at(1)?)?;
+		let ping_to = NodeEndpoint::from_rlp(&rlp.at(2)?)?;
 		let timestamp: u64 = rlp.val_at(3)?;
 		self.check_timestamp(timestamp)?;
 
 		let mut response = RlpStream::new_list(3);
-		dest.to_rlp_list(&mut response);
+		let pong_to = NodeEndpoint {
+			address: from.clone(),
+			udp_port: ping_from.udp_port
+		};
+		// Here the PONG's `To` field should be the node we are
+		// sending the request to
+		ping_to.to_rlp_list(&mut response);
+		// pong_to.to_rlp_list(&mut response);
 		response.append(&echo_hash);
 		append_expiration(&mut response);
 		self.send_packet(PACKET_PONG, from, &response.drain())?;
 
-		let entry = NodeEntry { id: *node, endpoint: source.clone() };
+		let entry = NodeEntry { id: *node_id, endpoint: pong_to.clone() };
 		if !entry.endpoint.is_valid() {
 			debug!(target: "discovery", "Got bad address: {:?}", entry);
+			return Ok(None);
 		} else if !self.is_allowed(&entry) {
 			debug!(target: "discovery", "Address not allowed: {:?}", entry);
-		} else {
-			self.add_node(entry.clone());
+			return Ok(None);
 		}
-
-		Ok(None)
+		Ok(self.update_node(entry.clone()))
```

Comment:
Why the change from `add_node` to `update_node`? According to [this](https://github.com/ethereum/devp2p/blob/master/discv4.md#endpoint-proof), the node should only be added to the routing table after we receive a PONG (not just a PING).

Reply:
To quote your link: _When a ping packet is received, the recipient should reply with a pong packet. It may also consider the sender for addition into the node table._

Reply:
IMO it's just another unnecessary round of Ping-Pong.

Reply:
Yes, `add_node` considers the sender for inclusion, but only after obtaining Endpoint Proof, as it is referred to in the doc.

I disagree that it is unnecessary. The purpose of PINGing the node is to ensure that it is listening at the network address it claims to be and participating in the protocol. When receiving a single UDP packet, the origin address can be spoofed, or the packet could be sent without the sender actually listening on the network. The linked document details the difficulty of verifying that a peer is honestly running a discovery server, but PINGing is a good heuristic.

Reply:
Note that Geth doesn't Ping back an received Ping, it just adds it to their NodeTable .

Reply:
I don't really know how devp2p works, but I think that it is useful to wait for the PONG, as nodes behind a NAT (which is probably not uncommon) may report an unreachable address.